### ISSUE TITLE
Can not access _ingest metadata from painless scripting

### ISSUE THREAD

**Issue by jakelandis** [elastic/elasticsearch#60470]
Each processor can add an "if" condition that runs a painless script to evaluate if the processor should be run. The script processor also executes a painless script. Currently the only information available to Painless (via ctx.) is the _source and metadata (i.e. _index) from the ingest document, not the _ingest metadata. The _ingest metadata has access to the ingestion time, pipeline used, etc. This information is only available via the mustache script templates which are not accessible via painless. 

It is arguably a bug that the templates (mustache script) and painless scripts are not aligned. For example:

```
DELETE foo

POST foo/_doc?pipeline=foo_pipeline
{
  "a" : 1,
  "_ingest" : "blah"
 
}

PUT _ingest/pipeline/foo_pipeline
{
  "processors": [
    {
      "set": {
        "field": "x",
        "value": "{{_ingest}}"
      }
    },
    {
      "script": {
        "source": "ctx.y = ctx._ingest"
      }
    }
  ]
}

GET foo/_search
```
results in
```
{
        "_index" : "foo",
        "_type" : "_doc",
        "_id" : "4Z0MKoEBL4S8u-TTDs0h",
        "_score" : 1.0,
        "_source" : {
          "a" : 1,
          "_ingest" : "blah",
          "x" : "{pipeline=foo_pipeline, timestamp=2022-06-03T14:50:41.970148170Z}",
          "y" : "blah"
        }
      }
```
If exposing Painless to _ingest meta data should be read only since the internal processing should be the only thing writing to the field. (all values are already read only via the if condition)

Perhaps the most impactful area is the for-each processor.  The for-each processor uses a special `_ingest` meta data to help keep track of which value you are working with. Since `_ingest` metadata is not exposed to painless for either the script process or the "if" condition it makes using painless with the for-each process not really possible since you can not pull the current value from the `_ingest.value`. 

We should consider exposing the `_ingest` data to painless for the script processor and "if" condition. The workaround is to for-go the for-each process and do the necessary looping and processing via a top level (i.e. not inside a for-each) script processor. Additionally using the set and remove processor to read data from `{{_ingest.value}}` and then set a temporary field in the main document (and remove at the end of the pipeline) may allow for some (awkward) usage of painless within a for-each.

**Comment by stu-elastic** [comment]
This is worth doing but needs to be prioritized, anyone who wants to pick it up, feel free.

**Comment by AaronYYC** [comment]
Might I help on this issue?
I've found map object sourceAndMetadata and ingestMetadata in IngestDocument.java,
and will try to figure out how these objects are being processed, is that a good starting point?

**Comment by AaronYYC** [comment]
Hi @stu-elastic, is this issue still active?

**Comment by jakelandis** [comment]
@AaronYYC - yes, all issues should be active if they are opened and if unassigned no one likely actively working on it. 

You are in the right section of the code base, the ingest document is what is available to each processor, but I don't have the specifics on the implementation (I haven't look too closely at how to resolve this).

**Comment by AaronYYC** [comment]
@jakelandis Is it possible to provide a test case that hits " The _ingest metadata is not exposed to painless for either the script process or the 'if' condition." ? 

 I've been searching for relative test cases in ConditionalProcesssorTests, ScriptProcessorTests, ForEachProcessorTests, bunch of YamlTests  and many others but still in doubt if I can forge a Test case that fits the requirement. 

Grateful if you can provide such test cases and expected results.

Something interesting I found in 
https://github.com/elastic/elasticsearch/blob/master/server/src/test/java/org/elasticsearch/ingest/IngestDocumentTests.java#L76-L87
Since IngestDocument can get _ingest thru _source._ingest, I assume that it's those being tracked working values needs to be exposed to script process or the "if" condition. 

Perhaps we need to add some functionalities to the processor's tracking mechanism?

**Comment by cyberpion-yotam** [comment]
Has this gone anywhere? 
I'm trying to conditionally append values from one array to another, and have to resort to using the "script" processor instead of the intuitive:
```
{
  "append": {
    "field": "dns.txt_answers",
    "value": "{{_ingest._value.data}}",
    "if": "{{_ingest._value.type}} == 'TXT'"
  }
}
```

**Comment by jakelandis** [comment]
The fix for this is also arguably a breaking change..but it likely a very rare occurrence.  If we implement this by adding _ingest to ctx, i.e. ctx._ingest, then it is possible for us to collide with documents that have a field named _ingest as a top most field. The following scenario would be considered breaking:
* A source document has a field called '_ingest' 
* The source document is processed via the ingest node and there is an existing Painless script processor that reads the `_ingest` field from the source document. 

After we fix this bug, that processor would now be reading the internal _ingest field, not the one provided by the document. Thus possibly breaking the script processor. The work around would be to rename the documents _ingest field to avoid the collision. This seems that it would be very rare, and the breaking change would well out-weighed by the benefits of having access to this field.

**Comment by andrewkroh** [comment]
> and the breaking change would be out-weighed by the benefits of having access to this field

I fully agree. 👍

**Comment by stu-elastic** [comment]
The collision between fields and metadata values in `ctx` is why we're adding the `Metadata` object.  Rather than exposing this via the `ctx` map, I'd recommend putting it in `Metadata`. See: https://github.com/elastic/elasticsearch/pull/87309

**Comment by joegallo** [comment]
> Additionally using the set and remove processor to read data from `{{_ingest.value}}` and then set a temporary field in the main document (and remove at the end of the pipeline) may allow for some (awkward) usage of painless within a for-each.

I ended up looking at this ticket again today, and while I'm not thrilled with what I'm about to type, I do want to record for posterity that the above mentioned workaround does work fine (and is indeed awkward)

```
PUT _ingest/pipeline/for_each_ingest_inner
{
  "processors": [
    {
      "set": {
        "description": "Work around lack of access to _ingest from conditionals and scripts",
        "field": "current",
        "copy_from": "_ingest._value"
      }
    },
    {
      "uppercase": {
        "description": "Actual work here, we want to only uppercase the values that start with f",        
        "field": "current",
        "if": "ctx.current.startsWith('f')"
      }
    },
    {
      "set": {
        "description": "Work around lack of access to _ingest from conditionals and scripts",
        "field": "_ingest._value",
        "copy_from": "current"
      }
    },
    {
      "remove": {
        "description": "Work around lack of access to _ingest from conditionals and scripts",        
        "field": "current"
      }
    }
  ]
}

PUT _ingest/pipeline/for_each_ingest
{
  "processors": [
    {
      "foreach" : {
        "field" : "values",
        "processor" : {
          "pipeline" : {
            "name" : "for_each_ingest_inner"
          }
        }
      }
    }
  ]
}

POST /_ingest/pipeline/for_each_ingest/_simulate?verbose
{
  "docs": [
    {
      "_index": "index",
      "_id": "id",
      "_source": {
        "values": ["foo", "bar", "baz"]
      }
    }
  ]
}
```

That is, I'd like to 'just' have a foreach around an uppercase with an if, but in the current implementation for scripts and conditionals doesn't make that possible, so I use some sets with copy_from to get around it, and then finally a remove to make sure I tidy up after myself.

None of this comment is meant to indicate that this workaround is *desirable*, I'm merely running with the observation that it's possible and showing exactly how to actually do it.

A nearby aside: yet another feature here that would be keen is if the `foreach` processor accepted a list of processors rather than just a single processor...

**Comment by stu-elastic** [comment]
> `IngestConditionalScript` extends neither `IngestScript` nor `WriteScript`, so it doesn't have the metadata() method, or at least that's my immediate read of things at a casual glance.

That is correct, I sketched out what that change would look like in a [branch](https://github.com/stu-elastic/elasticsearch/commit/7e028f9277146ab86bafe4856198f0098cbb7c2d).  That's just a start, it needs to enforce read-only behavior.

**Comment by m-adams** [comment]
I just came accross this as my document had an array in it.
I wanted to use a foreach ingest processor and then a small painless script processor to process each element but as the foreach processor passes the array element as _ingest._value it isn't accessible from painless.
I realise I can just do it all in painless but it would be nicer to support painless inside foreach in some fashion.

**Comment by rjernst** [comment]
re ingest conditional scripts, the fields api is being added to it in https://github.com/elastic/elasticsearch/pull/121914.

Coming back to the original issue, we already have other scripts, specifically AggregationScript, which has a very similar `_value`. We could add `_ingest` in the same way to IngestScript. Note however there is one potential issue:
we haven't limited variable names created by users, so `_ingest` could break scripts. Stuffing it into metadata doesn't quite seem right (it's not document metadata, it's more a special context variable added by the foreach processor). Another possibility is making the context variables we provide available with a different prefix that is specially reserved (eg `$ingest`).

If adding `_ingest` is something the data management team wants, they can do that (no changes in painless necessary). If we want to have special prefixes some work will be necessary in painless, but we'll need to split that off from here.

**Comment by jdconrad** [comment]
Closing this as it's unlikely to be prioritized.
